core/domain/state_domain.py

```diff
@@ -918,8 +942,933 @@ def get_all_outcomes(self) -> List[Outcome]:
+                        den['denominator'] ==
+                        rule_spec.inputs['f']['denominator']
+                    ):
+                        raise utils.ValidationError(
+                            f'Rule \'{rule_spec_index}\' from answer '
+                            f'group \'{ans_group_index}\' of '
+                            f'FractionInput interaction having '
+                            f'rule type HasFractionalPart'
+                            f'ExactlyEqualTo will '
+                            f'never be matched because it is '
+                            f'made redundant by the above rules'
+                        )
+
+                ranges.append(range_var)
+                matched_denominator_list.append(matched_denominator)
+
+    def _validate_number_with_units_input(self) -> None:
+        """Validates the NumberWithUnitsInput interaction.
+
+        Raises:
+            ValidationError. Duplicate rules are present.
+            ValidationError. The 'IsEqualTo' rule comes after 'IsEquivalentTo'
+                rule having same values.
+        """
+        number_with_units_rules = []
+        rule_spec_till_now: List[RuleSpecDict] = []
+
+        for ans_group_index, answer_group in enumerate(self.answer_groups):
+            for rule_spec_index, rule_spec in enumerate(
+                answer_group.rule_specs
+            ):
+                # Rule should not be duplicate.
+                if rule_spec.to_dict() in rule_spec_till_now:
+                    raise utils.ValidationError(
+                        f'The rule \'{rule_spec_index}\' of answer group '
+                        f'\'{ans_group_index}\' of NumberWithUnitsInput '
+                        f'interaction is already present.'
+                    )
+                rule_spec_till_now.append(rule_spec.to_dict())
+
+                # `IsEqualTo` rule should not come after `IsEquivalentTo` rule.
+                if rule_spec.rule_type == 'IsEquivalentTo':
+                    number_with_units_rules.append(rule_spec.inputs['f'])
+                if (
+                    rule_spec.rule_type == 'IsEqualTo' and
+                    rule_spec.inputs['f'] in number_with_units_rules
+                ):
+                    raise utils.ValidationError(
+                        f'The rule \'{rule_spec_index}\' of answer '
+                        f'group \'{ans_group_index}\' has '
+                        f'rule type equal is coming after '
+                        f'rule type equivalent having same value '
+                        f'in FractionInput interaction.'
+                    )
+
+    def _validate_choices_should_be_unique_and_empty(self) -> None:
+        """Validates that the answer choice should be unique and non empty
+
+        Raises:
+            ValidationError. Answer choice is empty.
+            ValidationError. Answer choice is duplicate.
+        """
+        choices = self.customization_args['choices'].value
```

Comment:
Can be handled via schema

Reply:
Sorry, can you explain it a bit more?

Reply:
Oh Sorry,
What I meant was if we introduce a schema validation of not having `empty choices` then the `default_value` will raise an error as in the default value of `MultipleChoice` or `ItemSelection` we introduce empty choice for the starter. More descriptive explanation would be something like this [part](https://github.com/oppia/oppia/blob/develop/extensions/interactions/DragAndDropSortInput/DragAndDropSortInput.py#L49)

Reply:
nvm, got it! So even if we are adding these validations in the state class the default values are invalid, right? (I think, that's fine i.e, some default values are invalid e.g, content_id of SubtitledHtml object).

> There will be an issue 

What will be the issue? Imean do you see any issue in the frontend?

Reply:
By issue I meant if we are introducing a validation and default value does not follow that then it will raise the error, right? I'm not completely sure about this if the error will be raised or not, want to confirm this part also with you.
I don't think there will be any issue with the frontend part.
Thanks

Reply:
>  I'm not completely sure about this if the error will be raised or not, want to confirm this part also with you.

I would suggest trying and seeing whether it fails (if it won't take hrs. to test it)!

Reply:
Done! It is raising an error.
```
FAIL: test_default_interactions_are_valid (extensions.interactions.base_test.InteractionUnitTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/lkbhitesh07/Desktop/oppia/oppia/extensions/interactions/base_test.py", line 686, in test_default_interactions_are_valid
    interaction.customization_arg_specs)
  File "/home/lkbhitesh07/Desktop/oppia/oppia/extensions/interactions/base_test.py", line 161, in _validate_customization_arg_specs
    ca_spec.default_value, ca_spec.schema))
  File "/home/lkbhitesh07/Desktop/oppia/oppia/core/schema_utils.py", line 218, in normalize_against_schema
    ) for item in obj
  File "/home/lkbhitesh07/Desktop/oppia/oppia/core/schema_utils.py", line 218, in <listcomp>
    ) for item in obj
  File "/home/lkbhitesh07/Desktop/oppia/oppia/core/schema_utils.py", line 295, in normalize_against_schema
    validator['id'], kwargs, normalized_obj))
AssertionError: Validation failed: has_subtitled_html_non_empty ({}) for object {'content_id': None, 'html': ''}

```

Reply:
So an existing test is failing, right? (And not the UI/UX)

Considering our default value isn't valid for some of the cases and we know it can we do something for default value such that it doesn't fail test. (Maybe have a flag that represents the default value isn't valid?)

Reply:
Done!